File: BookSearchObjc/BookCollectionObjc/Screens/UsingTableView/BookDetails/Controller/BookDetailsViewController.m
//
//  BookDetailsViewController.m
//  BookCollectionObjc
//
//  Created by Milton Palaguachi on 11/14/20.
//

#import "BookDetailsViewController.h"
#import "BookDetail.h"
#import "DatabaseManager.h"

@interface BookDetailsViewController ()

@property (nonatomic, strong) Book *book;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIImageView *bookImage;
@property (nonatomic, strong) NSMutableArray<BookDetail *> *details;
- (void)downloadImage:(NSURL * _Nonnull ) url;

@end

@implementation BookDetailsViewController
static DatabaseManager *shared;
- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        self.details = [NSMutableArray new];
    }
    return  self;
}
#pragma mark - UIView lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.tableView setEstimatedRowHeight:100.0];
    [self.tableView registerNib:[UINib nibWithNibName:@"BookDetailTableViewCell" bundle:nil] forCellReuseIdentifier:@"BookDetailTableViewCell"];
    [self setupTabBarItem];
    shared = [DatabaseManager sharedManager];
    [shared startDatabase];

}
- (void)setupTabBarItem {
    UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithTitle:@"Save" style: UIBarButtonItemStylePlain target:self action: @selector(saveBook:)];
    [self.navigationItem setRightBarButtonItem:item animated:YES];
}
- (void)saveBook:(id)sender {
//    DatabaseManager *dbManager= [DatabaseManager sharedManager];
    [shared insert: self.book];
    NSLog(@"save tapped");
    
}

- (void)configureWithBook: (Book *)book {
    self.book = book;
    if (book.title) {
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Title" description: book.title]];
    }
    
    if (book.subtitle) {
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Subtitle" description: book.subtitle]];
    }
   
    if (book.category) {
//        NSString *categories = [[book.category valueForKey:@"description"] componentsJoinedByString:@""];
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Caterogy" description: book.category]];
    }
    
    if (book.author) {
//        NSString *authors = [[book.author valueForKey:@"description"] componentsJoinedByString:@""];
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Authors" description: book.author]];
    }
    
    if (book.descriptionInfo) {
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Description" description: book.descriptionInfo]];
    }
    
    if (book.language) {
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Language" description: book.language]];
    }
    
    if (book.publisher) {
        [self.details addObject: [[BookDetail alloc] initWithTitle: @"Publisher" description: book.publisher]];
    }
   
//    [self.details addObject: [[BookDetail alloc] initWithTitle: @"Avg Rating" description: book.averageRating]];
//    [self.details addObject: [[BookDetail alloc] initWithTitle: @"Page Count" description: book.pageCount]];
    [self downloadImage: [[NSURL alloc] initWithString:book.thumbnail]];
    [self.tableView reloadData];
}

- (void)downloadImage: (NSURL * _Nonnull ) url {
    
    if (@available(iOS 13.0, *)) {
        dispatch_async(dispatch_get_global_queue(0,0), ^{
            NSData * data = [[NSData alloc] initWithContentsOfURL:url];
            if(!data) {
                NSLog(@"Error no data"); return;
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                self.bookImage.image = [UIImage imageWithData: data];
            });
        });
    } else {
        NSLog(@"// Fallback on earlier versions");
    }
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    BookDetailTableViewCell * cell = (BookDetailTableViewCell *)[self.tableView dequeueReusableCellWithIdentifier: @"BookDetailTableViewCell" forIndexPath: indexPath];
    if(!cell) {
        cell = [[BookDetailTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier: @"BookDetailTableViewCell"];
    }
    [cell configureWithCellViewModelProtocol: self.details[indexPath.row]];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.details.count;
}



@end

File: BookSearchSwift/BookCollection/Screens/BookDetails/Controller/BookDetailsViewController.swift
//
//  BookDetailsViewController.swift
//  BookCollection
//
//  Created by Milton Palaguachi on 11/16/20.
//  Copyright © 2020 Milton. All rights reserved.
//

import UIKit

class BookDetailsViewController: UIViewController {
    
    @IBOutlet private weak var bookImageView: UIImageView!
    @IBOutlet private weak var tableView: UITableView! {
        didSet {
            self.tableView.delegate = self
            self.tableView.dataSource = self
            self.tableView.tableFooterView = UIView()
            self.tableView.register(DetailTableViewCell.nib(), forCellReuseIdentifier: DetailTableViewCell.identifier)
        }
    }
        
    private var previewLink: String?
    var volumeInfo: VolumeInfoModel?
    typealias Detail = (type: DetailType, info: String)
    private var bookDetils: [Detail] = [] {
        didSet {
            self.tableView.reloadData()
        }
    }
    // MARK: - ViewDidLoad
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setProperties()
        self.setupDetailTableView()
        self.setupNavigationItem()
        self.downloadImage()
    }
    
    // MARK: - ViewDidAppear
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(true)
        self.setupLink()
    }
    
    // MARK: - Set Properties
    @objc func webSearch (_ sender: UIBarButtonItem) {
        guard let url = URL(string: previewLink ?? "") else { return }
        UIApplication.shared.open(url)
    }
    
    // MARK: - Setup Right BarButtonItem
    
    private func setupNavigationItem() {
        
        let safari = UIBarButtonItem(image: Constants.safariImage, style: .plain, target: self, action: #selector(webSearch(_:)))
        let done = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(doneAction(_:)))
        self.navigationItem.leftBarButtonItems = [done, safari]
    }
    
    public func addSaveBarItem() {
        let save = UIBarButtonItem(barButtonSystemItem: .save, target: self, action: #selector(saveVolumeToDadaBase(_:)))
        self.navigationItem.rightBarButtonItem = save
        
    }
    
    // MARK: - Download Book Image
    
    private func downloadImage() {
        guard let thumbNail = volumeInfo?.bookImageUrl, let url = URL(string: thumbNail) else { return }
        self.bookImageView.downloadImage(with: url)
    }
    
    @objc private func saveVolumeToDadaBase(_ sender: UIBarButtonItem) {
        if let volume = self.volumeInfo {
            do {
                try DBManager.shared.insert(volume: volume)
            } catch {
                print("error - \(error)")
            }
        }
    }
    
    @objc private func doneAction(_ sender: UIBarButtonItem) {
        self.dismiss(animated: true, completion: nil)
    }
    
    // MARK: - Setup Link for book
    
    private func setupLink() {
        if let previewLink = volumeInfo?.previewLink {
            self.previewLink = previewLink
        }
    }
    
    // MARK: - Append Data to dictArray
    
    private func setProperties() {
        
        guard let volume = volumeInfo else { return }
        if let title = volume.title, !title.isEmpty {
            self.bookDetils.append((.title, title))
        }
        if let subtitle = volume.subtitle, !subtitle.isEmpty {
            self.bookDetils.append((.subtitle, subtitle))
        }
        
        if let author = volume.authors, !author.isEmpty {
            self.bookDetils.append((.author, author))
        }
        
        if let publisher = volume.publisher, !publisher.isEmpty {
            self.bookDetils.append((.publisher, publisher ))
        }
        
        if let publishedDate = volume.publishedDate, !publishedDate.isEmpty {
            self.bookDetils.append((.publishedDate, publishedDate))
        }
        
        if let description = volume.description, !description.isEmpty {
            self.bookDetils.append((.description, description ))
        }
        
        if let count = volume.pageCount, count != 0 {
            self.bookDetils.append((.pageCount, "\(count)" ))
        }
        
        if let averageRating = volume.averageRating, averageRating != 0.0 {
            self.bookDetils.append((.averageRating, "\(averageRating)" ))
        }
        
        if let language = volume.language, !language.isEmpty {
            self.bookDetils.append((.language, language))
        }
    }
    
    // MARK: - TableView Setup
    
    private func setupDetailTableView() {
        self.tableView.rowHeight = UITableView.automaticDimension
        self.tableView.estimatedRowHeight = 50
        self.tableView.allowsSelection = false
    }
}

// MARK: - UITableViewDataSource And UITableViewDelegate

extension BookDetailsViewController: UITableViewDataSource, UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        self.bookDetils.count
    }
    
    // MARK: - CellForRowAt Method
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let detail = self.bookDetils[indexPath.row]
        let cellModel = CellModel( title: detail.type.rawValue, subtitle: detail.info)
        
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "DetailTableViewCell", for: indexPath) as? DetailTableViewCell else {
            fatalError("Unable to dequeue DetailTableViewCell")
        }
        cell.configure(model: cellModel)
        return cell
    }
}
